sdks/python/apache_beam/dataframe/frames.py

```diff
@@ -3655,6 +3655,40 @@ def shift(self, axis, freq, **kwargs):
   describe = _agg_method(pd.DataFrame, 'describe')
   max = _agg_method(pd.DataFrame, 'max')
   min = _agg_method(pd.DataFrame, 'min')
+
+  @frame_base.with_docs_from(pd.DataFrame)
+  @frame_base.args_to_kwargs(pd.DataFrame)
+  @frame_base.populate_defaults(pd.DataFrame)
+  def pivot(self, **kwargs):
+    columns = kwargs.get('columns', None)
+    selected_cols = self._expr.proxy()[columns]
+
+    if isinstance(selected_cols, pd.Series):
+      all_cols_are_categorical = isinstance(
+        selected_cols.dtype, pd.CategoricalDtype
+      )
+    else:
+      all_cols_are_categorical = all(
+        isinstance(c, pd.CategoricalDtype) for c in selected_cols.dtypes
+      )
+    if not all_cols_are_categorical:
+      raise frame_base.WontImplementError(
+          "pivot() of non-categorical type is not supported because "
+          "the type of the output column depends on the data. Please use "
+          "pd.CategoricalDtype with explicit categories.",
+          reason="non-deferred-columns")
+
+    proxy = pd.DataFrame(columns=self._expr.proxy().pivot(**kwargs).columns)
+
+    with expressions.allow_non_parallel_operations():
```

Comment:
I don't think it's a good idea to bypass the non-parallel operations here. The places we do that are cases where we know the data has already been reduced substantially (e.g. we just need to combine partial results for a global aggregation). In this case the input _could_ be very large, so we'd want the user to have to acknowledge that they want to use Singleton partitioning.

That being said, I think we should be able to do this without Singleton partitioning. What if we first move the `index` column(s) to the index in one expression (if set), then apply the `columns`/`values` part in a separate expression, with `requires=Index()`?